contracts/compressor/IntentCompressor.sol

```diff
@@ -0,0 +1,335 @@
+                    );
+                }
+
+                // Transfer tokens from funding source to vault using safe transfer
+                IERC20(token).safeTransferFrom(funder, vault, remainingAmount);
+            }
+        }
+    }
+
+    function _constructIntent(
+        EncodedIntent memory encodedIntent
+    ) internal view returns (Intent memory) {
+        Route memory route = _constructRoute(
+            _randomBytes32(),
+            encodedIntent.sourceChainIndex,
+            encodedIntent.destinationChainIndex,
+            encodedIntent.routeTokenIndex,
+            encodedIntent.routeAmount
+        );
+
+        Reward memory reward = _constructReward(
+            encodedIntent.rewardTokenIndex,
+            encodedIntent.rewardAmount,
+            encodedIntent.expiryDuration
+        );
+
+        return Intent({route: route, reward: reward});
+    }
+
+    function _constructReward(
+        uint8 rewardTokenIndex,
+        uint48 rewardAmount,
+        uint24 expiryDuration
+    ) internal view returns (Reward memory) {
+        TokenAmount memory rewardToken = TokenAmount({
+            token: getTokens()[rewardTokenIndex],
+            amount: rewardAmount
+        });
+
+        TokenAmount[] memory rewardTokens;
+        rewardTokens[0] = rewardToken;
+
+        return
+            Reward({
+                nativeValue: 0,
+                prover: PROVER,
+                creator: msg.sender,
+                tokens: rewardTokens,
+                deadline: block.timestamp + expiryDuration
+            });
+    }
+
+    function _constructRoute(
+        bytes32 salt,
+        uint8 sourceChainIndex,
+        uint8 destinationChainIndex,
+        uint8 routeTokenIndex,
+        uint256 routeAmount
+    ) internal view returns (Route memory) {
+        address routeTokenTarget = getTokens()[routeTokenIndex];
+
+        Call memory routeCallTransfer = Call({
+            target: routeTokenTarget,
+            value: 0,
+            data: abi.encodeCall(IERC20.transfer, (msg.sender, routeAmount))
+        });
+
+        TokenAmount memory routeToken = TokenAmount({
+            token: routeTokenTarget,
+            amount: routeAmount
+        });
+
+        TokenAmount[] memory routeTokens;
+        routeTokens[0] = routeToken;
+
+        Call[] memory routeCalls;
+        routeCalls[0] = routeCallTransfer;
+
+        return
+            Route({
+                inbox: address(INBOX),
+                salt: salt,
+                source: getChainIds()[sourceChainIndex],
+                destination: getChainIds()[destinationChainIndex],
+                calls: routeCalls,
+                tokens: routeTokens
+            });
+    }
+
+    // ======================== Private Functions ========================
+
+    function _randomBytes32() private view returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(block.timestamp, block.prevrandao, msg.sender)
```

Comment:
wanna add `block.coinbase` and maybe `tx.gasprice` into the mix?

Reply:
I think we need to come up with something better, this still doesn't allow to create 2 identical intents in the same block...

Reply:
Hmmm, interesting. Instead, we can use a much smaller salt instead of doing it this way. We still have 11 bytes free in the compressed payload. We can reduce the salt to 8 bytes and add it to the payload

Reply:
I'm doing this, thanks @re1ro 🙂

Reply:
How about we add `gasleft()` into the mix and it will produce different salts for the same intents within a batch?

Reply:
That's a great idea, but then it is a bit unpredictable getting the resulting intent hash.

Reply:
Isn't it the whole point of random value to be unpredictable? 😄 
`block.timestamp, block.prevrandao` aren't very predictable either...

Reply:
Basically we just need to make sure to have all different salts within the same batch, what we have now, it will output the same salt for the whole batch, is that right?